KAFKA-6104: Added unit tests for ClusterConnectionStates.

soenkeliebau opened the pull request:
Added a few unit tests for ClusterConnectionStates.

soenkeliebau committed:
KAFKA-6104: Added unit tests for ClusterConnectionStates.

tedyu commented:
Please fix indentation warnings.

tedyu commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -0,0 +1,125 @@
+package org.apache.kafka.clients;
```
Add license header

ijuma commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -0,0 +1,142 @@
+        assertTrue("Expected connection to be connecting after initializing connection attempt.", connectionStates.isConnecting(nodeId));
+        assertFalse("Expected connection not to be ready for requests during an ongoing connection attempt.", connectionStates.isReady(nodeId));
+        assertFalse("Expected connection not to be blocked during an ongoing connection attempt", connectionStates.isBlackedOut(nodeId, time.milliseconds()));
+        assertFalse("Expected connection to have no ready nodes.", connectionStates.hasReadyNodes());
```
I think most (all?) of these assertions are very obvious and the long text is just repeating the code. Does the text add value of is it just noise? Also, the lines should not be longer than the GitHub review window.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -0,0 +1,142 @@
+        long lowerLimit = Math.round(reconnectBackoffMsTest * (1 - reconnectBackoffJitter));
+        long upperLimit = Math.round(reconnectBackoffMsTest * (1 + reconnectBackoffJitter));
+        long currentBackoff = connectionStates.connectionDelay(nodeId, time.milliseconds());
+        assertTrue("Expected initial backoff after disconnect to be within 20% jitter of 'reconnect.backoff.ms' parameter.", lowerLimit <= currentBackoff && currentBackoff <= upperLimit);
```
If you use the `assertEquals` that takes a `double`, you can pass a `delta` value, which makes the code a lot more concise.

ijuma commented on clients/src/test/java/org/apache/kafka/clients/ClusterConnectionStatesTest.java:
```diff
@@ -0,0 +1,142 @@
+
+    @Test
+    public void testMaxReconnectBackoff() {
+        Long effectiveMaxReconnectBackoff = Math.round(reconnectBackoffMaxTest * (1 + reconnectBackoffJitter));
```
Seems like this should be a `long`.

ijuma reviewed (commented):
Thanks a lot for the tests! Just a few minor comments.

soenkeliebau commented:
@ijuma I've addressed all your comments. I agree on the assert messages, I guess I was just trying to be holier than the pope here :)
I've also added a testcase with more than one node to check that nothing gets confused in that scenario.

ijuma commented:
LGTM, merged to trunk and 1.0 branches after minor style fixes.
